test: headers sync timeout

stringintech opened the pull request:
When reviewing PR #32051 and considering which functional tests might need to be adapted/extended accordingly, I noticed there appears to be limited functional test coverage for header sync timeouts and thought it might be helpful to add one.

This test attempts to cover two scenarios:

1. **Normal peer timeout behavior:** When a peer fails to respond to initial getheaders requests within the timeout period, it should be disconnected and the node should attempt to sync headers from the next available peer.

2. **Noban peer behavior:** When a peer with noban privileges times out, it should remain connected while the node still attempts to sync headers from other peers.

i-am-yuvi commented:
Concept ACK

Great observation and thanks for adding. I was thinking if we could add this timeout test in [`p2p_initial_headers_sync.py`](https://github.com/bitcoin/bitcoin/blob/master/test/functional/p2p_initial_headers_sync.py) instead of creating another one.

stringintech commented:
> I was thinking if we could add this timeout test in [`p2p_initial_headers_sync.py`](https://github.com/bitcoin/bitcoin/blob/master/test/functional/p2p_initial_headers_sync.py) instead of creating another one.

@yuvicc Actually, I did not consider that. However, now that I am thinking about it, the two sub-tests in the new test file share some context that are not used by the test file you mentioned. So, I think it is cleaner to keep them separate, especially if it will be extended to test the addnode peers as well (like [here](https://github.com/stringintech/bitcoin/commit/02f4bec94db6db81b2885646312219ddabf82270#diff-8ba81fc6003c09eec6425cd72ae7384e982dde0a7a5ca12a9a9349a41879df95)). But I will be thinking more about it while waiting for other feedback.

mzumsande reviewed (commented):
Concept ACK.
Didn't read the above, but I also thought it might be better to add the additional tests to `p2p_initial_headers_sync.py` - after all, this is not a particularly long test.

stringintech commented on test/functional/p2p_initial_headers_sync_timeout.py:
```diff
@@ -0,0 +1,111 @@
+    def trigger_timeout(self):
+        self.log.info("Trigger the headers download timeout by advancing "
+                      "mock time")
+        best_header_time = self.node.getblockchaininfo()['time']
```
I used `getblockchaininfo()['time']` because we hadn't sent the node any headers yet, so I relied on the assumption that the best header should still be at genesis with the genesis block time (which I should have commented better).
But I agree with your suggestion as it is more direct and I have applied it - I appreciate your further feedback on it.

mzumsande replied:
I looked at this again, and I think the original approach is fine - I think I was confused by something in my original comment, sorry!
Just adding a comment that the genesis block time is the best header time in this case (because we didn't send any headers) should be sufficient - no need to send a block.

stringintech replied:
Thanks for looking into this further. I fixed the timing issue I mentioned earlier, but while reviewing it, I started to question how much value sending a new header/block actually adds to the test. I’ll revert it to use the genesis block time as the best header time and add the comment.

But does changing the order of the subtests actually matter, given that the `test_initial_headers_sync` subtest sends no actual headers?

stringintech replied:
(Just pushed a new version; I'll update the subtest order once this conversation wraps up. Thanks!)

mzumsande replied:
right, it just sends an empty headers message - kind of surprising that a test called `test_initial_headers_sync` doesn't send a single header. I think it's fine to leave the order as is.

stringintech commented:
16a62fa to bff80f5 - Addressed feedback by @mzumsande and @yuvicc; Thanks!

Prabhat1308 reviewed (commented):
Concept ACK . left a nit

mzumsande commented on test/functional/p2p_initial_headers_sync.py:
```diff
@@ -66,35 +92,137 @@ def run_test(self):
+        peer1.send_without_ping(msg_headers())
 
-        assert_equal(count, 1)
+        self.log.info(
```
I think it would be good to drop the formatting changes to code you don't change for easier review. Some linters such as flake8 will suggest a 80 character limit, but that is not usually followed here - there will be longer lines in almost every functional test (it's fine to respect in the added code if you want).

stringintech replied:
Done. Good to know the line length limit isn't strictly enforced everywhere since respecting it actually hurts readability in some places.

stringintech commented:
Drafted to address the issue explained [here](https://github.com/bitcoin/bitcoin/pull/32677#discussion_r2153287651) and the new comments.

stringintech committed:
test: headers sync timeout

mzumsande reviewed (commented):
Code Review ACK 8d257ed5937a47e1a60b7fe085f2e984eb60a956

maflcko reviewed (commented):
lgtm. Nice test for both branches

review ACK 8d257ed5937a47e1a60b7fe085f2e984eb60a956 👋

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 8d257ed5937a47e1a60b7fe085f2e984eb60a956 👋
Dx0uDH6sK4oEHrtvkvxApjRYlTLYj+EHHZzNAl++iaYjo/3lz0ow5P9Hx+JBjxTIAPOCUijhrwtWNKESdnxHDQ==
```

</details>

stringintech commented on test/functional/p2p_initial_headers_sync.py:
```diff
@@ -24,18 +27,30 @@
+
+def calculate_headers_timeout(best_header_time, current_time):
+    seconds_since_best_header = current_time - best_header_time
+    variable_timeout_us = (HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER_US *
+                           seconds_since_best_header / POW_TARGET_SPACING_SEC)
+    variable_timeout_sec = variable_timeout_us / 1_000_000
+    return int(current_time + HEADERS_DOWNLOAD_TIMEOUT_BASE_SEC + variable_timeout_sec)
```
I think the +1 offset would still be needed in a corner case when the input of the ceil function is already an integer. In this case, if +1 is removed, the timeout will not be triggered because the check `current_time > peer->m_headers_sync_timeout` ([here](https://github.com/bitcoin/bitcoin/blob/7d5a6d17398aae3cea6d63564d33e7e49da3c71c/src/net_processing.cpp#L5888)) would evaluate to `timeout > timeout` (false) instead of `timeout + 1 > timeout` (true).

stringintech replied:
Aside from the +1 offset though, maybe it is good to apply your suggestion and use the ceil function so that `calculate_headers_timeout` returns a time that is >= the real timeout calculated by core, ensuring it doesn't return a lower value due to precision loss.

stringintech commented:
[8d257ed](https://github.com/bitcoin/bitcoin/commit/8d257ed5937a47e1a60b7fe085f2e984eb60a956) to [1154618](https://github.com/bitcoin/bitcoin/commit/11546183c70def6c0aa539642fd1c9ada3d46840) ([compare](https://github.com/bitcoin/bitcoin/compare/8d257ed5937a47e1a60b7fe085f2e984eb60a956..11546183c70def6c0aa539642fd1c9ada3d46840)) - Addresses review comments from @maflcko and @l0rinc. Thanks!

stratospher reviewed (commented):
ACK 11546183c70def6c0aa539642fd1c9ada3d46840. nice test!
left some more style nits here - https://github.com/bitcoin/bitcoin/pull/32677#discussion_r2204080826

stratospher reviewed (approved):
reACK 61e800e7.
